gen/generate.go

```diff
@@ -150,105 +205,101 @@ func generateModule(m *compile.Module, i thriftPackageImporter, o *Options) erro
 	// for this Thrift file.
 	importPath, err := i.Package(m.ThriftPath)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	// packageOutDir is the directory whithin which all files and folders for
-	// this Thrift file will be written.
-	packageOutDir := filepath.Join(outDir, packageRelPath)
-
-	// Mapping of file names relative to $packageOutDir and their contents.
-	files := make(map[string]*bytes.Buffer)
+	// Mapping of file names relative to packageRelPath to their contents.
+	// Note that we need to return a mapping relative to o.OutputDir so we
+	// will prepend $packageRelPath/ to all these paths.
+	files := make(map[string][]byte)
 
 	if len(m.Constants) > 0 {
 		g := NewGenerator(i, importPath, packageName)
 
 		for _, constantName := range sortStringKeys(m.Constants) {
 			if err := Constant(g, m.Constants[constantName]); err != nil {
-				return err
+				return nil, err
 			}
 		}
 
 		buff := new(bytes.Buffer)
 		if err := g.Write(buff, token.NewFileSet()); err != nil {
-			return fmt.Errorf(
+			return nil, fmt.Errorf(
 				"could not generate constants for %q: %v", m.ThriftPath, err)
 		}
 
 		// TODO(abg): Verify no file collisions
-		files["constants.go"] = buff
+		files["constants.go"] = buff.Bytes()
 	}
 
 	if len(m.Types) > 0 {
 		g := NewGenerator(i, importPath, packageName)
 
 		for _, typeName := range sortStringKeys(m.Types) {
 			if err := TypeDefinition(g, m.Types[typeName]); err != nil {
-				return err
+				return nil, err
 			}
 		}
 
 		buff := new(bytes.Buffer)
 		if err := g.Write(buff, token.NewFileSet()); err != nil {
-			return fmt.Errorf(
+			return nil, fmt.Errorf(
 				"could not generate types for %q: %v", m.ThriftPath, err)
 		}
 
 		// TODO(abg): Verify no file collisions
-		files["types.go"] = buff
+		files["types.go"] = buff.Bytes()
 	}
 
 	if len(m.Services) > 0 {
 		for _, serviceName := range sortStringKeys(m.Services) {
 			service := m.Services[serviceName]
+
+			// If we called generateModule on something, that's a root
+			// service. We need plugins to generate code for them.
```

Comment:
It calls them on all modules by default, but only on one if `--no-recurse` was used. In that case, we want to generate code only for those services in that file, but we still need information about the services it depends on.